freqtrade/rpc/api_server/ws/proxy.py

```diff
@@ -0,0 +1,69 @@
+from typing import Any, Tuple, Union
+
+from fastapi import WebSocket as FastAPIWebSocket
+from websockets import WebSocketClientProtocol as WebSocket
+
+from freqtrade.rpc.api_server.ws.types import WebSocketType
+
+
+class WebSocketProxy:
+    """
+    WebSocketProxy object to bring the FastAPIWebSocket and websockets.WebSocketClientProtocol
+    under the same API
+    """
+
+    def __init__(self, websocket: WebSocketType):
+        self._websocket: Union[FastAPIWebSocket, WebSocket] = websocket
+
+    @property
+    def remote_addr(self) -> Tuple[Any, ...]:
+        if hasattr(self._websocket, "remote_address"):
+            return self._websocket.remote_address
+        elif hasattr(self._websocket, "client"):
+            return tuple(self._websocket.client)
+        return ("unknown", 0)
+
+    async def send(self, data):
+        """
+        Send data on the wrapped websocket
+        """
+        if isinstance(data, str):
+            data = data.encode()
+
+        if hasattr(self._websocket, "send_bytes"):
+            await self._websocket.send_bytes(data)
+        else:
+            await self._websocket.send(data)
+
+    async def recv(self):
+        """
+        Receive data on the wrapped websocket
+        """
+        if hasattr(self._websocket, "receive_bytes"):
+            return await self._websocket.receive_bytes()
+        else:
+            return await self._websocket.recv()
+
```

Comment:
I don't think we should enforce binary connections.
That's again something that the consumer/producer should negotiate (same as if they are using msgpack or json).


if i'm looking at a JS consumer - the default is string messages - which is currently not possible.
While we can convert it to binary - that's unnecessary - as the contained message is a string anyway (actually a string serialized object).

my current js code has to look like this:
```
utf8Encode.encode(
            JSON.stringify({
              type: 'subscribe',
              data: ['whitelist', 'entry_fill', 'exit_fill'],
            }),
          ),
```

the "encode" is however unnecessary - and will make debugging more difficult, as the network tab will no longer contain the messages in a readable format.

Reply:
Has been changed :+1: